server/src/test/java/org/elasticsearch/search/SearchHitTests.java

```diff
@@ -110,21 +124,34 @@ public static SearchHit createTestItem(boolean withOptionalInnerHits) {
         if (randomBoolean()) {
             hit.explanation(createExplanation(randomIntBetween(0, 5)));
         }
-        if (withOptionalInnerHits) {
+        if (isInnerHit == false && randomBoolean()) {
             int innerHitsSize = randomIntBetween(0, 3);
             if (innerHitsSize > 0) {
                 Map<String, SearchHits> innerHits = new HashMap<>(innerHitsSize);
                 for (int i = 0; i < innerHitsSize; i++) {
-                    innerHits.put(randomAlphaOfLength(5), SearchHitsTests.createTestItem());
+                    innerHits.put(randomAlphaOfLength(5),
+                        SearchHitsTests.createTestItem(xContentType, true));
                 }
                 hit.setInnerHits(innerHits);
             }
         }
-        if (randomBoolean()) {
+        /**
+         * The rest layer serialize the shard target iff:
+         *   - The search hit is not an inner hit.
+         *   - And {@link SearchHit#explanation} is not null.
+         * The rest serialization also looses the index uuid so we use
+         * {@link IndexMetaData#INDEX_UUID_NA_VALUE} as the uuid for the rest
+         * case.
+         */
+        if (isInnerHit == false
+                && (xContentType == null || hit.getExplanation() != null)
+                && randomBoolean()) {
```

Comment:
I think I understand now: you use xContentType null to figure out whether we are generating an item for transport or rest serialization?

Reply:
yes there is a comment above explaining the expectations.

Reply:
I've read the comment but it's not so clear what the null check is for, especially for people that haven't written this test. I wish that there would be another way to do this other than using null xContentType. Maybe that's where we should stop extending our randomizations and write a specific test for some very specific scenario? It would be clearer to me if we had a new test method that focused on shard target only, and tested only REST serialization. What do you think? Otherwise we may end up adding two different methods for generating randomized content (transport vs REST) but that sounds like overkill if only one test class has this issue.

Reply:
> Maybe that's where we should stop extending our randomizations and write a specific test for some very specific scenario?

I think we still want to randomize the tests for the rest serialization. This doesn't mean that we cannot have specific tests for some specific scenario but they should not replace the randomized tests imo. 

>  I wish that there would be another way to do this other than using null xContentType.

We can try to do it differently but the gist of this is that we have some objects that depends on the xContent type for the (de)serialization to be correct so it should be part of the randomization when we create these objects.

> Otherwise we may end up adding two different methods for generating randomized content (transport vs REST) but that sounds like overkill if only one test class has this issue.

We can also have two different test classes, one for the transport that extends AbstractStreamableTestCase and one for the rest layer that extends AbstractStreamableXContentTestCase. I don't think that's overkill if this ensures that we test
all possible serializations.

Reply:
I did not mean to suggest stopping using randomized tests. I meant that we should maybe stop extending the randomizations we already have to handle all the possible scenarios as it's proving complicated, hard to read and hard to maintain in the future. I am happy with what we have in your PR, having two different test classes would have been an option indeed but your PR has taken the single class approach which I am fine with too. I would just simplify the randomizations like we had them before handling shard target rest serialization, and add a test method that covers that scenario explicitly. I would find that clearer, easier to describe and to reason about.

Reply:
Ok thanks, I'll add a specific test for the shard target and remove the null xContentType check.